### PR TITLE
Fix unnecessary braces

### LINKED ISSUE #324: Inspection: unnessary braces in use statement.
For code like

```
use ra_db::input::{CrateId};
```

we should suggest removing the `{}`. The first step would be merely detecting such cases and producing a warning. The code for this is here:

https://github.com/rust-analyzer/rust-analyzer/blob/d77520fde3c953968beb09a3da80a0e7b17bbc04/crates/ra_editor/src/lib.rs#L92-L107

To check the file, one should taverse it using `.descendats()` method of the root node and check for a use tree with a single child.  A similar logic was recently implented [here](https://github.com/rust-analyzer/rust-analyzer/commit/67ac0a423f767ce8973348d5188e8baa6979d380)

Then one needs to add a `Severety` type & field to `Diagnostic`, so that we can distinguish between errors and warnings. Currently all diagnostis are errors, but this new one should be a warning. Then, fixing all usages of `Diagnostic` should basically lead to this being supported in the LSP.

### PR DESCRIPTION
**Pull request by gfreezy**
image

don't know why the underscore line is not covering the whole braces.

fix #324

### CODE REVIEW

File: crates/ra_analysis/src/imp.rs [hunk 1]
```diff
@@ -3,31 +3,32 @@ use std::{
...
 use hir::{
     self,
-    source_binder,
     FnSignatureInfo,
     Problem,
+    source_binder,
+};
+use ra_db::{FilesDatabase, SourceRoot, SourceRootId, SyntaxDatabase};
+use ra_editor::{self, FileSymbol, find_node_at_offset, LineIndex, LocalEdit, Severity};
+use ra_syntax::{
```
Comment (gfreezy) [inline crates/ra_analysis/src/imp.rs]: seems intellij-rust reorder the use statements. Should i revert it?

File: crates/ra_editor/src/typing.rs [hunk 2]
```diff
@@ -248,24 +254,13 @@ fn join_single_use_tree(edit: &mut TextEditBuilder, node: SyntaxNodeRef) -> Opti
...
-            if mem::replace(&mut res, Some(tree)).is_some() {
-                return None;
-            }
-        } else {
-            match child.kind() {
-                WHITESPACE | L_CURLY | R_CURLY | COMMA => (),
-                _ => return None,
-            }
-        }
+pub(crate) fn single_use_tree(tree_list: ast::UseTreeList) -> Option<ast::UseTree> {
```
Comment (gfreezy) [inline crates/ra_editor/src/typing.rs]: i rewrited this function. it seems easier for me.

File: crates/ra_editor/src/lib.rs [hunk 3]
```diff
@@ -124,11 +126,27 @@ fn check_unnecessary_braces_in_use_statement(file: &SourceFileNode) -> Vec<Diagn
...
+                let range = use_tree_list.syntax().range();
+                // use_tree_list always has one child, so we use unwrap directly here.
+                let to_replace = typing::single_use_tree(use_tree_list)
+                    .unwrap()
+                    .syntax()
+                    .text()
+                    .to_string();
+                let mut edit_builder = TextEditBuilder::new();
+                edit_builder.delete(range);
+                edit_builder.insert(range.start(), to_replace);
```
Comment (DJMcNab) [inline crates/ra_editor/src/lib.rs]: This doesn't seem to handle `self` nicely. It's very rare that this would come up, but it could. See #316.

Comment (gfreezy) [reply crates/ra_editor/src/lib.rs]: `use a::b::{self}` should be fixed as `use a::b` instead of `use a::b::self`?

**Comment by DJMcNab** [comment]
Yes, exactly that.

File: crates/ra_analysis/src/imp.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_analysis/src/imp.rs]: No, that’s fine

**Comment by matklad** [comment]
lgtm!

bors r+

File: crates/ra_editor/src/lib.rs (hunk 3 above)
Comment (matklad) [reply crates/ra_editor/src/lib.rs]: Yeah, this can be handled in a separate PR

**Pull request merged**